crates/turbopack-ecmascript/src/path_visitor.rs

```diff
@@ -12,77 +12,91 @@ use crate::code_gen::VisitorFactory;
 
 pub type AstPath = Vec<AstParentKind>;
 
-pub struct ApplyVisitors<'a> {
+pub struct ApplyVisitors<'a, 'b> {
     /// `VisitMut` should be shallow. In other words, it should not visit
     /// children of the node.
-    visitors: HashMap<AstParentKind, Vec<(&'a AstPath, &'a dyn VisitorFactory)>>,
+    visitors: Cow<'b, [(&'a AstPath, &'a dyn VisitorFactory)]>,
 
     index: usize,
 }
 
-impl<'a> ApplyVisitors<'a> {
-    pub fn new(visitors: Vec<(&'a AstPath, &'a dyn VisitorFactory)>) -> Self {
-        let mut map = HashMap::<AstParentKind, Vec<(&'a AstPath, &'a dyn VisitorFactory)>>::new();
-        for (path, visitor) in visitors {
-            if let Some(span) = path.first() {
-                map.entry(*span).or_default().push((path, visitor));
-            }
-        }
+/// Do two binary searches to find the sub-slice that has `path[index] == kind`.
+/// Returns None if no item matches that. `visitors` need to be sorted by path.
+fn find_range<'a, 'b>(
+    visitors: &'b [(&'a AstPath, &'a dyn VisitorFactory)],
+    kind: &AstParentKind,
+    index: usize,
+) -> Option<&'b [(&'a AstPath, &'a dyn VisitorFactory)]> {
+    // Precondition: visitors is never empty
+    let start = if visitors.first().unwrap().0[index] >= *kind {
+        // Fast path: It's likely that the whole range is selected
+        0
+    } else {
+        visitors.partition_point(|(path, _)| path[index] < *kind)
```

Comment:
Is there a guarantee that `path[index]` isn't going to panic? The `visitor` paths could be any length, right?

Reply:
That's within the invariant. I'll document that.